### PR TITLE
i#6662 regdeps: Invariant checks for filetype and cpuid

### PR DESCRIPTION
**Pull request by edeiana**
Public traces (i.e., OFFLINE_FILE_TYPE_ARCH_REGDEPS traces) have a new
constraint: TRACE_MARKER_TYPE_CPU_ID values must all be -1 as we don't
want to expose a CPU schedule. We add this constraint as invariant in the
`invariant_checker`.

We also add an invariant for the trace file type, which should not contain any
OFFLINE_FILE_TYPE_ARCH_* information except for
OFFLINE_FILE_TYPE_ARCH_REGDEPS.

Issue #6662

### CODE REVIEW

**Comment by edeiana** [comment]
`code_api|tool.drcachesim.coherence` failed on ci-x86 / x86-64 (pull_request):
```
2024-11-28T03:48:11.0247145Z 319: 
2024-11-28T03:48:11.0247551Z 319:      Finished computing current solution distance in mode 0.
2024-11-28T03:48:11.0248094Z 319:      Mode changed to 0.
2024-11-28T03:48:11.0248447Z 319: 
2024-11-28T03:48:11.0248778Z 319:      Started iteration 2 of the computation...
2024-11-28T03:48:11.0249227Z 319: 
2024-11-28T03:48:11.0249626Z 319:      Finished computing current solution distance in mode 0.
2024-11-28T03:48:11.0250159Z 319:      Mode changed to 0.
2024-11-28T03:48:11.0250507Z 320: 
2024-11-28T03:48:11.0250836Z 320:      Started iteration 1 of the computation...
2024-11-28T03:48:11.0251401Z 319: 
2024-11-28T03:48:11.0251747Z 319:      Started iteration 3 of the computation...
2024-11-28T03:48:11.0252191Z 319: 
2024-11-28T03:48:11.0252588Z 319:      Finished computing current solution distance in mode 0.
2024-11-28T03:48:11.0253126Z 319:      Mode changed to 0.
2024-11-28T03:48:11.0253464Z 319: 
2024-11-28T03:48:11.0253712Z 319: 
2024-11-28T03:48:11.0254022Z 319:      The Jacobi Method For AX=B .........DONE
2024-11-28T03:48:11.0254507Z 319:      Total Number Of iterations   :  3
2024-11-28T03:48:11.0255015Z 319:     ...................................................................
2024-11-28T03:48:11.0255626Z 319: Invalid trace entry type thread_exit (23) before a bundle
2024-11-28T03:48:11.0257713Z 319: drmemtrace_launcher: /home/runner/work/dynamorio/dynamorio/clients/drcachesim/reader/reader.cpp:262: virtual bool dynamorio::drmemtrace::reader_t::process_input_entry(): Assertion `type_is_instr(cur_ref_.instr.type) || cur_ref_.instr.type == TRACE_TYPE_INSTR_NO_FETCH' failed.
2024-11-28T03:48:11.0259583Z 320: 
2024-11-28T03:48:11.0259996Z 320:      Finished computing current solution distance in mode 0.
2024-11-28T03:48:11.0260538Z 320:      Mode changed to 0.
2024-11-28T03:48:11.0261378Z 320/515 Test #319: code_api|tool.drcachesim.coherence .................................***Failed  Required regular expression not found. Regex=[^
```
xref #4673

**Comment by derekbruening** [comment]
> code_api|tool.drcachesim.coherence failed on ci-x86 / x86-64 (pull_request):

You can see that this failure is not an output mismatch; it is a fatal assert:

> Invalid trace entry type thread_exit (23) before a bundle

That's #3320

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 1]
```diff
@@ -1638,6 +1638,36 @@ invariant_checker_t::check_regdeps_invariants(per_shard_t *shard, const memref_t
...
+                    shard->file_type_,
+                "TRACE_MARKER_TYPE_FILETYPE and shard file type are not the same");
+            // Check that the file type does not contain any architecture specific
+            // information except OFFLINE_FILE_TYPE_ARCH_REGDEPS.
+            report_if_false(shard,
+                            !TESTANY(OFFLINE_FILE_TYPE_ARCH_AARCH64, shard->file_type_),
+                            "OFFLINE_FILE_TYPE_ARCH_REGDEPS traces cannot contain "
+                            "OFFLINE_FILE_TYPE_ARCH_AARCH64");
+            report_if_false(shard,
+                            !TESTANY(OFFLINE_FILE_TYPE_ARCH_ARM32, shard->file_type_),
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:1654]: Could combine all of these (TESTANY checks for any of the bits): the message would be less specific but that doesn't seem so bad?

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 2]
```diff
@@ -43,6 +43,8 @@
 #include "../tools/invariant_checker.h"
```
Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:43]: Title:

> i#6662: invariant_checker update

For multi-commit issues we use a very short title after the issue number (see https://dynamorio.org/page_code_reviews.html#sec_commit_messages), sometimes with ", part N": so something like `i#6662 regdeps` for the prefix.  Plus I think you could fit what the update is about: so I would suggest this title: `i#6662 regdeps: Invariant checks for filetype and cpuid`.

**Review by derekbruening — APPROVED** [review]

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 2 above)
Comment (edeiana) [reply clients/drcachesim/tests/invariant_checker_test.cpp:43]: Agreed, much better description of just "update". Done.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 1 above)
Comment (edeiana) [reply clients/drcachesim/tools/invariant_checker.cpp]: Agreed, we don't really need a specific error message for every arch.
Collapsed into one check/error message using `OFFLINE_FILE_TYPE_ARCH_ALL`.
Done.

**Pull request merged**